src/hardware/sky-wallet.go

```diff
@@ -0,0 +1,446 @@
+	if txn != nil {
+		_, isSkycoinTxn := txn.(skytypes.SkycoinTxn)
+		if !isSkycoinTxn {
+			return false, nil
+		}
+	}
+	_, isSkycoinWlt := wlt.(skytypes.SkycoinWallet)
+	if !isSkycoinWlt {
+		// FIXME i18n
+		return false, errors.New("a non Skycoin wallet received in ReadyForTxn")
+	}
+	return hwMatchWallet(sw, wlt), nil
+}
+
+// NewSkyWallet create a new sky wallet instance
+func NewSkyWallet(dev skyWallet.Devicer, buttonAckHandler func(dev skywallet.Devicer, prvMsg wire.Message, outsLen int) (wire.Message, error)) *SkyWallet {
+	return &SkyWallet{
+		dev: dev,
+		handleButtonAckSequence: buttonAckHandler,
+	}
+}
+
+func getAllIndexesFromTxn(txn core.Transaction) []int {
+	inputs := txn.GetInputs()
+	indexes := make([]int, len(inputs), cap(inputs))
+	for i := 0; i < len(inputs); i++ {
+		indexes[i] = i
+	}
+	return indexes
+}
+
+func findValInSlice(val int, arr []int) bool {
+	for i := 0; i < len(arr); i++ {
+		if arr[i] == val {
+			return true
+		}
+	}
+	return false
+}
+
+func getInputs(txn coin.Transaction, indexes []int) (inputs []*messages.SkycoinTransactionInput, err error) {
+	for idx, input := range txn.In {
+		var transactionInput messages.SkycoinTransactionInput
+		transactionInput.HashIn = proto.String(input.String())
+		if findValInSlice(idx, indexes) {
+			transactionInput.Index = proto.Uint32(uint32(idx))
+		}
+		inputs = append(inputs, &transactionInput)
+	}
+	return inputs, nil
+}
+
+func getOutputs(tr coin.Transaction) (inputs []*messages.SkycoinTransactionOutput, err error) {
+	for _, output := range tr.Out {
+		var transactionOutput messages.SkycoinTransactionOutput
+		transactionOutput.Address = proto.String(output.Address.String())
+		transactionOutput.Coin = proto.Uint64(output.Coins)
+		transactionOutput.Hour = proto.Uint64(output.Hours)
+		// FIXME: should be possible to send the address index
+		// find index to check if it is determined as a owned address
+		//	transactionOutput.AddressIndex = proto.Uint32(uint32(addressIndex[i]))
+		// check out hw implementation
+		//addrIndex, err := strconv.Atoi(s[i])
+		//if err != nil {
+		//      logrus.WithField("str_addr_index", s[i]).Error("unable to get integer from string")
+		//      return nil, errors.New("error getting address index as integer")
+		//}
+		//if addrIndex < 0 {
+		//      logrus.WithField("addr_index", addrIndex).Error("addrIndex should be greater than 0")
+		//      return nil, errors.New("addrIndex should be greater than 0")
+		// }
+		//transactionOutput.AddressIndex = proto.Uint32(uint32(addrIndex))
+		inputs = append(inputs, &transactionOutput)
+	}
+	return
+}
+
+func readableTxn2Transaction(txn skytypes.ReadableTxn) (*coin.Transaction, error) {
+	cTxn, err := txn.ToCreatedTransaction()
+	if err != nil {
+		return nil, err
+	}
+	skyTxn, err := cTxn.ToTransaction()
+	if err != nil {
+		return nil, err
+	}
+	return skyTxn, nil
+}
+
+func rawTxn2Transaction (txn *skycoin.SkycoinUninjectedTransaction) (*coin.Transaction, error) {
```

Comment:
Expand argument type scope . All instances of `SkycoinTxn` implement `EncodeSkycoinTransaction()`